docs(BulkWrite): adds documentation to BulkWrite

daprahamian opened the pull request:
## Description

Adds documentation for multiple methods and properties
associated with bulk writes. Also marks certain methods as
internal.

Fixes NODE-1800
Fixes NODE-1938

**What changed?**

Added more documentation to the BulkOperationBase. Includes, but is not limited to:
- Updating language
- Adding examples for how to add various operations to the bulk op
- Correcting the jsdoc for getters such that the description and type actually show up
- Marking certain methods as internal by adding an `@ignore` tag

**Are there any files to ignore?**
- `lib/bulk/ordered.js` and `lib/bulk/unordered.js` have trivial changes.

kvwalker commented on lib/bulk/common.js:
```diff
@@ -55,60 +55,69 @@ class Batch {
 
   /**
-   * @return {boolean} ok Did bulk operation correctly execute
+   * Is true if the bulk operation correctly execute
```
Might be worded better as `Returns true if the bulk operation correctly executes`

mbroadst commented on lib/bulk/common.js:
```diff
@@ -55,60 +55,69 @@ class Batch {
+  /**
+   * Create a new BulkWriteResult instance (INTERNAL TYPE, do not instantiate directly)
+   *
+   * @class
```
I believe `@class` is superfluous if you are using ES6 classes

mbroadst commented on lib/bulk/common.js:
```diff
@@ -220,7 +229,7 @@ class BulkWriteResult {
 
   /**
-   * @return {BulkWriteResult} a BulkWriteResult instance
+   * @return {object}
```
No docs here?

mbroadst replied:
SGTM

mbroadst commented on lib/bulk/common.js:
```diff
@@ -242,25 +251,30 @@ class BulkWriteResult {
+   * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
+   *
+   * @class
+   * @return {WriteConcernError} a WriteConcernError instance
```
we don't document return values for most constructors, should we be doing that here?

mbroadst commented on lib/bulk/common.js:
```diff
@@ -242,25 +251,30 @@ class BulkWriteResult {
  */
 class WriteConcernError {
+  /**
+   * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
```
nit: we have traditionally called out notes in bold, like:
```
Create a new WriteConcernError

**NOTE:** Internal type, do not instantiate directly
```
I don't know if there's another jsdoc tag for this

mbroadst commented on lib/bulk/common.js:
```diff
@@ -531,15 +551,17 @@ function handleMongoWriteConcernError(batch, bulkResult, err, callback) {
+  /**
+   * Creates a new BulkWriteError
+   *
+   * @class
```
no need for `@class`

mbroadst commented on lib/bulk/common.js:
```diff
@@ -531,15 +551,17 @@ function handleMongoWriteConcernError(batch, bulkResult, err, callback) {
- * @param {BulkWriteResult} result The result of the bulk write operation
- * @return {BulkWriteError} A BulkWriteError instance
- * @extends {MongoError}
+ * @classdesc An error indicating an unsuccessful Bulk Write
```
I also think `@classdesc` is superfluous

mbroadst commented on lib/bulk/common.js:
```diff
@@ -552,24 +574,25 @@ class BulkWriteError extends MongoError {
 
   /**
-   * Add a single update document to the bulk operation
+   * Add a multiple update operation to the bulk operation
```
why not just "Adds an update operation to the bulk operation"?

kvwalker commented on lib/bulk/common.js:
```diff
@@ -55,60 +55,69 @@ class Batch {
 
   /**
-   * @return {boolean} ok Did bulk operation correctly execute
+   * Evaluates to true if the bulk operation correctly execute
```
`execute` should be `executes` (or `executed`?)

kvwalker commented on lib/bulk/common.js:
```diff
@@ -282,39 +295,44 @@ class WriteConcernError {
 
   /**
-   * @return {string} errmsg Write concern error message.
+   * Write  error message.
```
I think there's an extra space here between Write and error

daprahamian commented:
@dandv I have corrected the class description to reference `delete(One)?` instead of the `remove` functions. I've also added an example for `replaceOne`, and added an `updateOne` example using a pipeline to leverage previous queried values (only available in 4.2).
